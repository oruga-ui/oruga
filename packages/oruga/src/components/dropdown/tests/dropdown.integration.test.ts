import { afterEach, describe, expect, test } from "vitest";
import { enableAutoUnmount, mount } from "@vue/test-utils";

import DropdownExample from "./DropdownExample.vue";
import ODropdownItem from "@/components/dropdown/DropdownItem.vue";
import ODropdown from "@/components/dropdown/Dropdown.vue";

describe("ODropdown Integration", () => {
    enableAutoUnmount(afterEach);

    test("is called", () => {
        const values = ["A", "B", "C"];
        const wrapper = mount(DropdownExample, { props: { values } });
        expect(wrapper.attributes("data-oruga")).toBe("dropdown");
        expect(!!wrapper.vm).toBeTruthy();

        const items = wrapper.findAllComponents(ODropdownItem);
        expect(items.length).toBe(values.length);
        values.forEach((value, idx) =>
            expect(items.at(idx).text()).toBe(value),
        );
    });

    test("react accordingly when new item is selected", async () => {
        const values = ["A", "B", "C"];
        const wrapper = mount(DropdownExample, {
            props: { values, value: values[0] },
        });

        const items = wrapper.findAll(".o-drop__item");
        expect(items.length).toBe(values.length);
        expect(items[0].classes("o-drop__item--active")).toBeTruthy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        await items[2].trigger("click");

        expect(items[0].classes("o-drop__item--active")).toBeFalsy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeTruthy();

        const dropdown = wrapper.findComponent(ODropdown);
        expect(dropdown.emitted("update:modelValue")).toHaveLength(1);
        expect(dropdown.emitted("change")).toHaveLength(1);
        expect(dropdown.emitted("change")[0][0]).toBe(values[2]);
        expect(dropdown.emitted("close")).toHaveLength(1);
    });

    test("react accordingly when same item is selected", async () => {
        const values = ["A", "B", "C"];
        const wrapper = mount(DropdownExample, {
            props: { values, value: values[0] },
        });

        const items = wrapper.findAll(".o-drop__item");
        expect(items.length).toBe(values.length);
        expect(items[0].classes("o-drop__item--active")).toBeTruthy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        await items[0].trigger("click");

        expect(items[0].classes("o-drop__item--active")).toBeTruthy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        const dropdown = wrapper.findComponent(ODropdown);
        expect(dropdown.emitted("update:modelValue")).toBeUndefined();
        expect(dropdown.emitted("change")).toBeUndefined();
        expect(dropdown.emitted("close")).toHaveLength(1);
    });

    test("react accordingly when an item is selected with multiple prop", async () => {
        const values = ["A", "B", "C"];
        const wrapper = mount(DropdownExample, {
            props: { values, multiple: true },
        });

        const items = wrapper.findAll(".o-drop__item");
        expect(items.length).toBe(values.length);
        expect(items[0].classes("o-drop__item--active")).toBeFalsy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        await items[0].trigger("click");

        expect(items[0].classes("o-drop__item--active")).toBeTruthy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        const dropdown = wrapper.findComponent(ODropdown);
        expect(dropdown.emitted("update:modelValue")).toHaveLength(1);
        expect(dropdown.emitted("change")).toHaveLength(1);
        expect(dropdown.emitted("change")[0][0]).toHaveLength(1);
        expect(dropdown.emitted("change")[0][0]).toContain(values[0]);
        expect(dropdown.emitted("close")).toBeUndefined();

        await items[2].trigger("click");

        expect(items[0].classes("o-drop__item--active")).toBeTruthy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeTruthy();

        expect(dropdown.emitted("update:modelValue")).toHaveLength(2);
        expect(dropdown.emitted("change")).toHaveLength(2);
        expect(dropdown.emitted("change")[1][0]).toHaveLength(2);
        expect(dropdown.emitted("change")[1][0]).toContain(values[0]);
        expect(dropdown.emitted("change")[1][0]).toContain(values[2]);
        expect(dropdown.emitted("close")).toBeUndefined();

        await items[0].trigger("click");
        expect(items[0].classes("o-drop__item--active")).toBeFalsy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeTruthy();

        expect(dropdown.emitted("update:modelValue")).toHaveLength(3);
        expect(dropdown.emitted("change")).toHaveLength(3);
        expect(dropdown.emitted("change")[2][0]).toHaveLength(1);
        expect(dropdown.emitted("change")[2][0]).toContain(values[2]);
        expect(dropdown.emitted("close")).toBeUndefined();
    });

    test("react accordingly when item is selected without closeable", async () => {
        const values = ["A", "B", "C"];
        const wrapper = mount(DropdownExample, {
            props: { values, closeable: false },
        });

        const items = wrapper.findAll(".o-drop__item");
        expect(items.length).toBe(values.length);
        expect(items[0].classes("o-drop__item--active")).toBeFalsy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        await items[0].trigger("click");

        expect(items[0].classes("o-drop__item--active")).toBeTruthy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        const dropdown = wrapper.findComponent(ODropdown);
        expect(dropdown.emitted("update:modelValue")).toHaveLength(1);
        expect(dropdown.emitted("change")).toHaveLength(1);
        expect(dropdown.emitted("close")).toBeUndefined();
    });

    test("react accordingly when is disabled", async () => {
        const values = ["A", "B", "C"];
        const wrapper = mount(DropdownExample, {
            props: { values, disabled: true },
        });

        const items = wrapper.findAll(".o-drop__item");
        expect(items.length).toBe(values.length);
        expect(items[0].classes("o-drop__item--active")).toBeFalsy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        await items[0].trigger("click");

        expect(items[0].classes("o-drop__item--active")).toBeFalsy();
        expect(items[1].classes("o-drop__item--active")).toBeFalsy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        const dropdown = wrapper.findComponent(ODropdown);
        expect(dropdown.classes("o-drop--disabled")).toBeTruthy();
        expect(dropdown.emitted("update:modelValue")).toBeUndefined();
        expect(dropdown.emitted("change")).toBeUndefined();
        expect(dropdown.emitted("close")).toBeUndefined();
    });

    test("react accordingly when is using objects values", async () => {
        const values = [{ label: "1" }, { label: "2" }, { label: "3" }];
        const wrapper = mount(DropdownExample, {
            props: { values, field: "label" },
        });

        const items = wrapper.findAll(".o-drop__item");
        expect(items.length).toBe(3);

        items.forEach((item, index) =>
            expect(item.text()).toEqual(values[index].label),
        );

        const item = items[1];
        await item.trigger("click");
        expect(items[0].classes("o-drop__item--active")).toBeFalsy();
        expect(items[1].classes("o-drop__item--active")).toBeTruthy();
        expect(items[2].classes("o-drop__item--active")).toBeFalsy();

        const dropdown = wrapper.findComponent(ODropdown);
        expect(dropdown.emitted("update:modelValue")).toHaveLength(1);
        expect(dropdown.emitted("update:modelValue")[0][0]).toStrictEqual(
            values[1],
        );
        expect(dropdown.emitted("change")).toHaveLength(1);
        expect(dropdown.emitted("change")[0][0]).toStrictEqual(values[1]);
        expect(dropdown.emitted("close")).toHaveLength(1);
    });
});
