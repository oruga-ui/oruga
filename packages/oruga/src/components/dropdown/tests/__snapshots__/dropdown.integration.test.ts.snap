// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dropdown integration tests > is called 1`] = `
"<div data-oruga="dropdown" class="o-drop o-drop--position-bottom-left">
  <div tabindex="0" class="o-drop__trigger" aria-haspopup="true">
    <!--
                @slot Override the trigger element, default is label prop
                @binding {boolean} active - dropdown active state
            --><button class="trigger">trigger</button>
  </div>
  <!--teleport start-->
  <transition-stub name="fade" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
  <transition-stub name="fade" appear="false" persisted="true" css="true">
    <div class="o-drop__menu o-drop__menu--bottom-left" role="list" aria-hidden="true" aria-modal="true" style="display: none;">
      <!--
                        @slot Place dropdown items here
                        @binding {boolean} active - dropdown active state
                        @binding {boolean} toggle - toggle active state function
                    -->
      <div class="o-drop__item o-drop__item--clickable text-break" role="listitem" tabindex="0" data-oruga="dropdown-item">
        <!--
            @slot Override the label, default is label prop 
        -->A
      </div>
      <div class="o-drop__item o-drop__item--clickable text-break" role="listitem" tabindex="0" data-oruga="dropdown-item">
        <!--
            @slot Override the label, default is label prop 
        -->B
      </div>
      <div class="o-drop__item o-drop__item--clickable text-break" role="listitem" tabindex="0" data-oruga="dropdown-item">
        <!--
            @slot Override the label, default is label prop 
        -->C
      </div>
    </div>
  </transition-stub>
  <!--teleport end-->
</div>"
`;
